app/code/core/Mage/CatalogRule/Model/Action/Index/Refresh.php

```diff
@@ -115,10 +115,7 @@ protected function _setResource(Mage_Core_Model_Resource_Db_Abstract $resource)
     public function execute()
     {
         $this->_app->dispatchEvent('catalogrule_before_apply', array('resource' => $this->_resource));
-
-        /** @var Mage_Core_Model_Date $coreDate */
-        $coreDate  = $this->_factory->getModel('core/date');
-        $timestamp = $coreDate->gmtTimestamp('Today');
+        $timestamp = Mage::app()->getLocale()->date(null, null, null,true)->get(Zend_Date::TIMESTAMP);
```

Comment:
```suggestion
        /** @var Mage_Core_Model_Date $coreDate */
        $coreDate  = $this->_factory->getModel('core/date');
        $timestamp = $coreDate->gmtTimestamp();
```

`Mage::app()->getLocale()->date(null, null, null,true)->get(Zend_Date::TIMESTAMP);` is a roundabout way to get the GMT timestamp, it goes to locale to get local time, then convert it to GMT.

However, I haven't tested if GMT timestamp fixes the issue.

Reply:
that's the solution that all the article promoted :-)

Reply:
Take a look: https://github.com/OpenMage/magento-lts/blob/3811f8ba3a3feb8906ef5899ac8ac4df39784bf5/app/code/core/Mage/Core/Model/Date.php#L142-L145

It returns GMT timestamp right away. I prefer this because:

1. Least change to original code.
2. The function name is self-explanatory that the result is a GMT timestamp, whereas the "accepted" code is not so clear.

I think GMT timestamp is the correct fix. See https://github.com/OpenMage/magento-lts/blob/3811f8ba3a3feb8906ef5899ac8ac4df39784bf5/app/code/core/Mage/CatalogRule/Model/Action/Index/Refresh.php#L623-L642

It uses `$timestamp` to form the query, and the DB stores datetime in GMT.

Reply:
but the original code that wasn't working had `$timestamp = $coreDate->gmtTimestamp('Today');` isn't it the same or am I missing something?

Reply:
No, it's not the same. You can test it easily:

```php
/** @var Mage_Core_Model_Date $coreDate */
$coreDate  = Mage::getSingleton('core/date');
$ts0 = $coreDate->gmtTimestamp('Today');
$ts1 = $coreDate->gmtTimestamp();
$ts2 = Mage::app()->getLocale()->date(null, null, null,true)->get(Zend_Date::TIMESTAMP);
Zend_Debug::dump([
    'gmt' => Varien_Date::now(),
    'ts0' => $ts0 . ' ' . date($format, $ts0),
    'ts1' => $ts1 . ' ' . date($format, $ts1),
    'ts2' => $ts2 . ' ' . date($format, $ts2),
]);
```

Output:
```
 array(4) {
  ["gmt"] => string(19) "2022-06-06 08:56:48"
  ["ts0"] => string(30) "1654444800 2022-06-05 16:00:00"
  ["ts1"] => string(30) "1654505808 2022-06-06 08:56:48"
  ["ts2"] => string(30) "1654505808 2022-06-06 08:56:48"
}
```

Reply:
switched to $coreDate->gmtTimestamp() ;-)